lib/ret/media_search.ex

```diff
@@ -0,0 +1,118 @@
+defmodule Ret.MediaSearchQuery do
+  @enforce_keys [:source]
+  defstruct [:source, :user, :filter, :q, page: 1]
+end
+
+defmodule Ret.MediaSearchResult do
+  @enforce_keys [:meta, :entries]
+  defstruct [:meta, :entries]
+end
+
+defmodule Ret.MediaSearchResultMeta do
+  @enforce_keys [:source, :page, :page_size, :total_pages, :total_entries]
+  defstruct [:source, :page, :page_size, :total_pages, :total_entries]
+end
+
+defmodule Ret.MediaSearch do
+  import Ret.HttpUtils
+  import Ecto.Query
+
+  alias Ret.{Repo, OwnedFile, SceneListing}
+
+  @page_size 24
+
+  def search(%Ret.MediaSearchQuery{source: "scene_listings", page: page, filter: "featured", q: query}) do
+    scene_listing_search(page, query, "featured", asc: :order)
+  end
+
+  def search(%Ret.MediaSearchQuery{source: "scene_listings", page: page, filter: filter, q: query}) do
+    scene_listing_search(page, query, filter)
+  end
+
+  def search(%Ret.MediaSearchQuery{source: "sketchfab", user: user}) do
+    with api_key when is_binary(api_key) <- resolver_config(:sketchfab_api_key) do
+      res =
+        "https://api.sketchfab.com/v3/search?type=models&downloadable=true&user=#{user}"
+        |> retry_get_until_success([{"Authorization", "Token #{api_key}"}])
+
+      case res do
+        :error ->
+          :error
+
+        res ->
+          res
+          |> Map.get(:body)
+          |> Poison.decode!()
+          |> Map.get("results")
+          |> Enum.map(&sketchfab_api_result_to_entry/1)
+      end
+    else
+      _ -> %{}
+    end
+  end
+
+  defp scene_listing_search(page, query, filter, order \\ [desc: :updated_at]) do
+    SceneListing
+    |> join(:inner, [l], s in assoc(l, :scene))
+    |> where([l, s], l.state == ^"active" and s.state == ^"active" and s.allow_promotion == ^true)
+    |> add_query_to_listing_search_query(query)
+    |> add_tag_to_listing_search_query(filter)
+    |> preload([:screenshot_owned_file, :model_owned_file, :scene_owned_file])
+    |> order_by(^order)
+    |> Repo.paginate(%{page: page, page_size: @page_size})
+    |> result_for_scene_listing_page()
+  end
+
+  def add_query_to_listing_search_query(query, nil), do: query
+  def add_query_to_listing_search_query(query, q), do: query |> where([l, s], ilike(l.name, ^"%#{q}%"))
```

Comment:
Will something in the ORM/Postgres do sanitization here? If not this looks like a potential SQL injection.

Reply:
Same for the tag query below, I think

Reply:
Also, I think these can be `defp`

Reply:
the ilike is sanitized for quote escaping but i'll check the tag one. however it's still a risk if/when this table gets too big because it's possible to construct some gnarly like clauses here (because you can inject `%`'s

Reply:
ah yes bc we are using `?` variable replacement w `fragment` there should be no injection afaik

Reply:
also the solution to the `ilike` one is to eventually migrate this into a proper search engine (postgresql has one built in which is probably where i'd start)

Reply:
Ah, you're right, `ilike` is already escaped. 👍